File: agents/java-agent/src/main/java/com/hmdev/messaging/agent/api/impl/MessagingChannelApi.java
package com.hmdev.messaging.agent.api.impl;


import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.hmdev.messaging.common.HttpClientResult;
import com.hmdev.messaging.common.HttpClient;
import com.hmdev.messaging.common.data.*;
import com.hmdev.messaging.common.security.MySecurity;
import com.hmdev.messaging.common.security.PemIO;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.io.ByteArrayInputStream;
import java.security.PublicKey;
import javax.crypto.Cipher;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.hmdev.messaging.agent.api.ConnectionChannelApi;

import java.net.URL;


public class MessagingChannelApi implements ConnectionChannelApi {
    private static final Logger logger = LoggerFactory.getLogger(MessagingChannelApi.class);

    private final static String PUBLIC_KEY = "public_key.php";

    // polling timeout in seconds
    private static final int POLLING_TIMEOUT = 40;
    private static final int DEFAULT_UDP_PORT = 9999;

    // Keep this set to false. A public key may be needed in the future, but HTTPS is sufficient for now.
    @Setter
    private boolean usePublicKey = false;

    private final HttpClient client;
    private final ObjectMapper objectMapper;

    // UDP helper
    private final UdpClient udpClient;

    public MessagingChannelApi(String remoteUrl, String developerApiKey) {
        this.client = new HttpClient(remoteUrl);
        this.objectMapper = new ObjectMapper();

        // If a developer API key was provided by the caller, attach it as a default header so all requests include X-Api-Key
        if (developerApiKey != null && !developerApiKey.isBlank()) {
            this.client.setDefaultHeader("X-Api-Key", developerApiKey);
        }

        // Load default developer API key from environment or system property if provided and set as default header
        // NOTE: Developer API key is no longer loaded from environment inside this class.
        // Agents or factories should call `setDeveloperApiKeyId(...)` when creating this API instance
        // if they want to attach an X-Api-Key header to requests.

        String host = "localhost";
        int udpPort = DEFAULT_UDP_PORT;
        try {
            URL url = new URL(remoteUrl);
            if (url.getHost() != null && !url.getHost().isEmpty()) {
                host = url.getHost();
            }
            if (url.getPort() > 0) {
                // If remoteUrl explicitly contains a udpPort, use it as the UDP udpPort (convenience)
                udpPort = url.getPort();
            }
        } catch (Exception e) {
            logger.warn("Unable to parse remoteUrl host/udpPort ({}}), defaulting to localhost:{}", remoteUrl, udpPort);
        }

        // Allow overriding UDP udpPort via system property or environment variable
        // -Dmessaging.udp.udpPort=XXXX or env MESSAGING_UDP_PORT
        try {
            String sysProp = System.getProperty("messaging.udp.udpPort");
            String envVal = System.getenv("MESSAGING_UDP_PORT");
            String chosen = (sysProp != null && !sysProp.isBlank()) ? sysProp : envVal;
            if (chosen != null && !chosen.isBlank()) {
                int overridePort = Integer.parseInt(chosen.trim());
                if (overridePort > 0 && overridePort <= 65535) {
                    udpPort = overridePort;
                    logger.info("Using UDP udpPort override: {}", udpPort);
                } else {
                    logger.warn("Ignoring invalid UDP udpPort override: {}", chosen);
                }
            }
        } catch (NumberFormatException nfe) {
            logger.warn("Invalid UDP udpPort override value; must be an integer: {}", nfe.getMessage());
        }

        this.udpClient = new UdpClient(host, udpPort, this.objectMapper);
    }

    @Override
    public ConnectResponse connect(String channelName, String channelPassword, String agentName)  {
        return connect(channelName, channelPassword, agentName, null, null);
    }

    @Override
    public ConnectResponse connect(String channelName, String channelPassword, String agentName, String sessionId)
    {
        return connect(channelName, channelPassword, agentName, sessionId, null);
    }

    @Override
    public ConnectResponse connect(String channelName, String channelPassword, String agentName, String sessionId,
                                   String channelId)  {
        return connect(channelName, channelPassword, agentName, sessionId, channelId, false);
    }

    @Override
    public ConnectResponse connect(String channelName, String channelPassword, String agentName, String sessionId,
                                   String channelId, boolean enableWebrtcRelay)  {
        return connect(channelName, channelPassword, agentName, sessionId, channelId, enableWebrtcRelay, "private");
    }

    @Override
    public ConnectResponse connect(String channelName, String channelPassword, String agentName, String sessionId,
                                   String channelId, boolean enableWebrtcRelay, String apiKeyScope)  {
        try {
            if (usePublicKey) {
                HttpClientResult publicKeyResponse = this.getPublicKey();

                if (publicKeyResponse.isHttpOk()) {
                    throw new Exception("Unable to get the public key");
                }
                PublicKey publicKey = PemIO.readPublicKey(new ByteArrayInputStream(publicKeyResponse.getData().getBytes()));
                Cipher pubKeyEncryptor = Cipher.getInstance("RSA");
                pubKeyEncryptor.init(Cipher.ENCRYPT_MODE, publicKey);
                client.setPublicKeyEncryptor(pubKeyEncryptor);
            }

            boolean hasChannelLogin = (channelName != null && channelPassword != null);
            String passwordHash = null;
            if (hasChannelLogin) {
                passwordHash = MySecurity.hash(channelPassword,
                        MySecurity.deriveChannelSecret(channelName, channelPassword));
            }

            if (channelId == null) {
                if (hasChannelLogin) {
                    // Create channel on server using channelName and passwordHash (protected password)
                    channelId = createChannel(channelName, passwordHash);
                }
                else
                {
                    throw new RuntimeException("Missing channelId or channelName+channelPassword for connect operation");
                }
            }

            // Build connect request: prefer channelId if known, but include name/password fields for compatibility
            ConnectRequest connectRequest = ConnectRequest.builder()
                    .channelId(channelId)
                    .channelName(channelName)
                    .channelPassword(passwordHash)
                    .agentName(agentName)
                    .agentContext(createAgentMetadata())
                    .build();
            connectRequest.setSessionId(sessionId);
            connectRequest.setEnableWebrtcRelay(enableWebrtcRelay);
            connectRequest.setApiKeyScope(apiKeyScope != null ? apiKeyScope : "private");

            HttpClientResult httpClientResult = this.client.request(HttpClient.RequestMethod.POST, getActionUrl("connect"), connectRequest);

            if (httpClientResult != null && httpClientResult.isHttpOk()) {
                // Controller wraps connect response in { status: 'success', data: { ... } }
                return objectMapper.readValue(
                        httpClientResult.dataAsJsonObject().optJSONObject("data").toString(), ConnectResponse.class);
            }
        } catch (Exception e) {
            logger.error("Exception caught in connect operation {}", e.getLocalizedMessage());
        }
        logger.debug("Unable to connect to the channel");
        return new ConnectResponse();
    }

    @Override
    public ConnectResponse connect(Map<String, Object> config) {
        String channelName = (String) config.get("channelName");
        String channelPassword = (String) config.get("channelPassword");
        String agentName = (String) config.get("agentName");
        String sessionId = (String) config.get("sessionId");
        String channelId = (String) config.get("channelId");
        boolean enableWebrtcRelay = config.getOrDefault("enableWebrtcRelay", false) instanceof Boolean ?
                (Boolean) config.get("enableWebrtcRelay") : false;
        String apiKeyScope = (String) config.getOrDefault("apiKeyScope", "private");

        return connect(channelName, channelPassword, agentName, sessionId, channelId, enableWebrtcRelay, apiKeyScope);
    }

    @Override
    public ConnectResponse connectWithChannelId(String agentName, String channelId, String sessionId) {
        return this.connect(null, null, agentName, sessionId, channelId, false);
    }

    @Override
    public ConnectResponse connectWithChannelId(String agentName, String channelId, String sessionId, boolean enableWebrtcRelay) {
        return this.connect(null, null, agentName, sessionId, channelId, enableWebrtcRelay);
    }

    // New helper: create channel on the server and return channelId if available
    private String createChannel(String name, String password) {
        try {
            HttpClientResult res = this.client.request(HttpClient.RequestMethod.POST,
                    getActionUrl("create-channel"), new CreateChannelRequest(name, password));

            if (res != null && res.isHttpOk()) {
                // todo: use dto instead of manual parsing
                return res.dataAsJsonObject().optJSONObject("data").optString("channelId", null);
            }
        } catch (Exception e) {
            logger.warn("create-channel failed: {}", e.getMessage());
        }
        return null;
    }


    @Override
    public EventMessageResult receive(String sessionId, ReceiveConfig receiveConfig) {
        EventMessageResult eventMessageResult = new EventMessageResult(new ArrayList<>(), null, null);
        try {
            MessageReceiveRequest messageReceiveRequest = new MessageReceiveRequest();
            messageReceiveRequest.setSessionId(sessionId);
            // Ensure we send JSON with fields globalOffset, localOffset, limit in correct order
            messageReceiveRequest.setReceiveConfig(new ReceiveConfig(receiveConfig.getGlobalOffset(),
                    receiveConfig.getLocalOffset(), receiveConfig.getLimit()));

            HttpClientResult httpClientResult = this.client.request(HttpClient.RequestMethod.POST, getActionUrl("pull"),
                    messageReceiveRequest, POLLING_TIMEOUT * 1000);


            if (httpClientResult.isHttpOk())
            {
                return objectMapper.readValue(httpClientResult.dataAsJsonObject().optJSONObject("data").toString(),
                        EventMessageResult.class);
            }
        } catch (Exception e) {
            logger.error("Exception caught in receive operation {}", e.getLocalizedMessage());
        }
        logger.debug("Unable to receive messages");
        return eventMessageResult;
    }

    @Override
    public List<AgentInfo> getActiveAgents(String sessionId) {
        try {
            SessionRequest sessionRequest = new SessionRequest(sessionId);

            HttpClientResult httpClientResult = this.client.request(HttpClient.RequestMethod.POST, getActionUrl("list-agents"),
                    sessionRequest);

            if (httpClientResult.isHttpOk()) {
                return objectMapper.readValue(httpClientResult.dataAsJsonObject().optJSONArray("data").toString(),
                        new TypeReference<>() {
                        });
            }

        } catch (Exception e) {
            logger.error("Exception for getActiveAgents operation: {}", e.getMessage());
        }
        return new ArrayList<>();
    }

    @Override
    public boolean send(EventMessage.EventType eventType, String msg, String destination, String sessionId, boolean encrypted) {

        try {
            EventMessageRequest eventMessageRequest = new EventMessageRequest();

            eventMessageRequest.setSessionId(sessionId);
            eventMessageRequest.setType(eventType);
            eventMessageRequest.setTo(destination);
            eventMessageRequest.setEncrypted(encrypted);
            eventMessageRequest.setContent(msg);

            return this.client.request(HttpClient.RequestMethod.POST,
                    getActionUrl("push"), eventMessageRequest).isHttpOk();
        }
        catch (Exception e) {
            logger.error("Exception for send operation: {}", e.getMessage());
        }

        return false;
    }

    @Override
    public boolean disconnect(String sessionId) {

        try {
            this.udpClient.close();
        } catch (Exception exception) {
            logger.debug("Error while closing udpClient: {}", exception.getMessage());
        }

        try {
            SessionRequest sessionRequest = new SessionRequest(sessionId);

            HttpClientResult response = this.client.request(HttpClient.RequestMethod.POST, getActionUrl("disconnect"),
                    sessionRequest);

            this.client.closeAll();

            return response.isHttpOk();
        }
        catch (Exception exception)
        {
            logger.error("Exception for disconnect operation: {}", exception.getMessage());
        }

        return false;
    }

    @Override
    public boolean udpPush(String msg, String destination, String sessionId) {
        try {
            EventMessageRequest eventMessageRequest = new EventMessageRequest();
            eventMessageRequest.setSessionId(sessionId);
            eventMessageRequest.setType(EventMessage.EventType.CHAT_TEXT);
            eventMessageRequest.setTo(destination);
            eventMessageRequest.setEncrypted(false);
            eventMessageRequest.setContent(msg);

            // Build UDP envelope using UdpEnvelope model
            UdpEnvelope envelope = new UdpEnvelope("push", objectMapper.valueToTree(eventMessageRequest));

            return this.udpClient.send(envelope);

        } catch (Exception e) {
            logger.error("Exception for udpPush operation: {}", e.getLocalizedMessage());
            return false;
        }
    }

    @Override
    public EventMessageResult udpPull(String sessionId, ReceiveConfig receiveConfig) {
        EventMessageResult eventMessageResult = new EventMessageResult(new ArrayList<>(), null, null);
        try {
            MessageReceiveRequest messageReceiveRequest = new MessageReceiveRequest();
            messageReceiveRequest.setSessionId(sessionId);
            messageReceiveRequest.setReceiveConfig(new ReceiveConfig(receiveConfig.getGlobalOffset(),
                    receiveConfig.getLocalOffset(), receiveConfig.getLimit()));

            // Build UDP envelope using UdpEnvelope model
            UdpEnvelope envelope = new UdpEnvelope("pull", objectMapper.valueToTree(messageReceiveRequest));

            // Send via UDP and wait for response (3 second timeout)
            JsonNode response = this.udpClient.sendAndWait(envelope, 3000);

            if (response != null) {
                // UDP response format: { status: "ok", result: { status: "success", data: {...} } }
                if (response.has("status") && "ok".equals(response.get("status").asText())) {
                    JsonNode result = response.get("result");
                    if (result != null && result.has("status") && "success".equals(result.get("status").asText())) {
                        JsonNode data = result.get("data");
                        if (data != null) {
                            return objectMapper.treeToValue(data, EventMessageResult.class);
                        }
                    }
                }
            }
        } catch (Exception e) {
            logger.error("Exception caught in udpPull operation {}", e.getLocalizedMessage());
        }
        return eventMessageResult;
    }

    public List<AgentInfo> getSystemAgents(String sessionId) {
        List<AgentInfo> agents = new ArrayList<>();
        try {
            SessionRequest sessionRequest = new SessionRequest(sessionId);
            HttpClientResult httpClientResult = this.client.request(HttpClient.RequestMethod.POST, getActionUrl("list-system-agents"), sessionRequest);
            if (httpClientResult.isHttpOk() && httpClientResult.dataAsJsonObject().optJSONArray("data") != null) {
                return objectMapper.readValue(httpClientResult.dataAsJsonObject().optJSONArray("data").toString(), new TypeReference<>() {});
            }
        } catch (Exception e) {
            logger.error("Exception for getSystemAgents operation: {}", e.getMessage());
        }
        return agents;
    }

    private HttpClientResult getPublicKey() {
        return this.client.request(HttpClient.RequestMethod.GET, PUBLIC_KEY, null);
    }

    private String getActionUrl(String action) {
        // Align exactly with kafka-service controller URLs
        return String.format("/%s", action);
    }

    /**
     * Create agent metadata map for connect requests
     * @return Map with agentType and descriptor
     */
    private Map<String, String> createAgentMetadata() {
        Map<String, String> metadata = new HashMap<>();
        metadata.put("agentType", "JAVA-AGENT");
        metadata.put("descriptor", MessagingChannelApi.class.getName());
        return metadata;
    }

}

File: agents/cpp-agent/src/messaging_channel_api.cpp
#include "hmdev/messaging/api/messaging_channel_api.h"
#include "hmdev/messaging/agent/security.h"
#include "hmdev/messaging/util/utils.h"
#include <stdexcept>
#include <iostream>

namespace hmdev {
namespace messaging {

MessagingChannelApi::MessagingChannelApi(const std::string& remoteUrl,
                                        const std::string& developerApiKey)
    : usePublicKey_(false), defaultPollSource_("AUTO") {

    // Create HTTP client
    httpClient_ = std::make_unique<HttpClient>(remoteUrl);

    // Set developer API key if provided
    if (!developerApiKey.empty()) {
        httpClient_->setDefaultHeader("X-Api-Key", developerApiKey);
    }

    // Parse URL to get host for UDP
    std::string host = "localhost";
    int udpPort = DEFAULT_UDP_PORT;

    std::string parsedHost;
    int parsedPort;
    if (Utils::parseUrl(remoteUrl, parsedHost, parsedPort)) {
        if (!parsedHost.empty()) {
            host = parsedHost;
        }
        if (parsedPort > 0) {
            udpPort = parsedPort;
        }
    }

    // Allow overriding UDP port via environment variable
    std::string envUdpPort = Utils::getEnv("MESSAGING_UDP_PORT");
    if (!envUdpPort.empty()) {
        try {
            int overridePort = std::stoi(envUdpPort);
            if (overridePort > 0 && overridePort <= 65535) {
                udpPort = overridePort;
            }
        } catch (const std::exception& e) {
            // Ignore invalid port
        }
    }

    // Create UDP client
    udpClient_ = std::make_unique<UdpClient>(host, udpPort);
}

MessagingChannelApi::~MessagingChannelApi() {
    // Unique pointers will auto-cleanup
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName) {
    return connect(channelName, channelPassword, agentName, "", "");
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName,
                                            const std::string& sessionId) {
    return connect(channelName, channelPassword, agentName, sessionId, "");
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName,
                                            const std::string& sessionId,
                                            const std::string& channelId) {
    return connect(channelName, channelPassword, agentName, sessionId, channelId, false);
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName,
                                            const std::string& sessionId,
                                            const std::string& channelId,
                                            bool enableWebrtcRelay) {
    return connect(channelName, channelPassword, agentName, sessionId, channelId, enableWebrtcRelay, "private");
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName,
                                            const std::string& sessionId,
                                            const std::string& channelId,
                                            bool enableWebrtcRelay,
                                            const std::string& apiKeyScope) {
    return connect(channelName, channelPassword, agentName, sessionId, channelId, enableWebrtcRelay, apiKeyScope, "AUTO");
}

ConnectResponse MessagingChannelApi::connect(const std::string& channelName,
                                            const std::string& channelPassword,
                                            const std::string& agentName,
                                            const std::string& sessionId,
                                            const std::string& channelId,
                                            bool enableWebrtcRelay,
                                            const std::string& apiKeyScope,
                                            const std::string& pollSource) {
    // Store default poll source for receive operations
    defaultPollSource_ = pollSource.empty() ? "AUTO" : pollSource;
        if (finalChannelId.empty()) {
            if (hasChannelLogin) {
                // Create channel on server
                finalChannelId = createChannel(channelName, passwordHash);
            } else {
                throw std::runtime_error("Missing channelId or channelName+channelPassword for connect operation");
            }
        }

        // Build connect request
        ConnectRequest connectRequest;
        connectRequest.channelId = finalChannelId;
        connectRequest.channelName = channelName;
        connectRequest.channelPassword = passwordHash;
        connectRequest.agentName = agentName;
        connectRequest.sessionId = sessionId;
        connectRequest.agentContext = createAgentMetadata();
        connectRequest.enableWebrtcRelay = enableWebrtcRelay;

        // Send connect request
        HttpClientResult result = httpClient_->post(getActionUrl("connect"),
                                                     connectRequest.toJson(),
                                                     POLLING_TIMEOUT_MS);

        if (result.isHttpOk()) {
            json responseJson = result.dataAsJson();
            if (responseJson.contains("data")) {
                return ConnectResponse::fromJson(responseJson["data"]);
        connectRequest.apiKeyScope = apiKeyScope.empty() ? "private" : apiKeyScope;
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in connect operation: " << e.what() << std::endl;
    }

    return ConnectResponse();
}

ConnectResponse MessagingChannelApi::connectWithChannelId(const std::string& agentName,
                                                          const std::string& channelId,
                                                          const std::string& sessionId) {
    return connect("", "", agentName, sessionId, channelId, false);
}

ConnectResponse MessagingChannelApi::connectWithChannelId(const std::string& agentName,
                                                          const std::string& channelId,
                                                          const std::string& sessionId,
                                                          bool enableWebrtcRelay) {
    return connect("", "", agentName, sessionId, channelId, enableWebrtcRelay);
ConnectResponse MessagingChannelApi::connect(const std::map<std::string, std::string>& config) {
    std::string channelName = config.count("channelName") ? config.at("channelName") : "";
    std::string channelPassword = config.count("channelPassword") ? config.at("channelPassword") : "";
    std::string agentName = config.count("agentName") ? config.at("agentName") : "";
    std::string sessionId = config.count("sessionId") ? config.at("sessionId") : "";
    std::string channelId = config.count("channelId") ? config.at("channelId") : "";
    bool enableWebrtcRelay = config.count("enableWebrtcRelay") && config.at("enableWebrtcRelay") == "true";
    std::string apiKeyScope = config.count("apiKeyScope") ? config.at("apiKeyScope") : "private";
    std::string pollSource = config.count("pollSource") ? config.at("pollSource") : "AUTO";

    return connect(channelName, channelPassword, agentName, sessionId, channelId, enableWebrtcRelay, apiKeyScope, pollSource);
}

                return responseJson["data"]["channelId"].get<std::string>();
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "create-channel failed: " << e.what() << std::endl;
    }

    return "";
}

EventMessageResult MessagingChannelApi::receive(const std::string& sessionId,
                                               const ReceiveConfig& config) {
    EventMessageResult result;

    try {
        MessageReceiveRequest request;
        request.sessionId = sessionId;

        // Apply default poll source if not specified in config
        ReceiveConfig effectiveConfig = config;
        if (effectiveConfig.pollSource.empty()) {
            effectiveConfig.pollSource = defaultPollSource_;
        }
        request.receiveConfig = effectiveConfig;

        HttpClientResult httpResult = httpClient_->post(getActionUrl("pull"),
                                                        request.toJson(),
                                                        POLLING_TIMEOUT_MS);

        if (httpResult.isHttpOk()) {
            json responseJson = httpResult.dataAsJson();
            if (responseJson.contains("data")) {
                return EventMessageResult::fromJson(responseJson["data"]);
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in receive operation: " << e.what() << std::endl;
    }

    return result;
}

std::vector<AgentInfo> MessagingChannelApi::getActiveAgents(const std::string& sessionId) {
    std::vector<AgentInfo> agents;

    try {
        SessionRequest request(sessionId);
        HttpClientResult result = httpClient_->post(getActionUrl("list-agents"),
                                                     request.toJson());

        if (result.isHttpOk()) {
            json responseJson = result.dataAsJson();
            if (responseJson.contains("data") && responseJson["data"].is_array()) {
                for (const auto& agentJson : responseJson["data"]) {
                    agents.push_back(AgentInfo::fromJson(agentJson));
                }
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in getActiveAgents: " << e.what() << std::endl;
    }

    return agents;
}

std::vector<AgentInfo> MessagingChannelApi::getSystemAgents(const std::string& sessionId) {
    std::vector<AgentInfo> agents;

    try {
        SessionRequest request(sessionId);
        HttpClientResult result = httpClient_->post(getActionUrl("list-system-agents"),
                                                     request.toJson());

        if (result.isHttpOk()) {
            json responseJson = result.dataAsJson();
            if (responseJson.contains("data") && responseJson["data"].is_array()) {
                for (const auto& agentJson : responseJson["data"]) {
                    agents.push_back(AgentInfo::fromJson(agentJson));
                }
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in getSystemAgents: " << e.what() << std::endl;
    }

    return agents;
}

bool MessagingChannelApi::send(EventType eventType,
                               const std::string& message,
                               const std::string& destination,
                               const std::string& sessionId,
                               bool encrypted) {
    try {
        EventMessageRequest request;
        request.sessionId = sessionId;
        request.type = eventType;
        request.to = destination;
        request.content = message;
        request.encrypted = encrypted;

        HttpClientResult result = httpClient_->post(getActionUrl("push"),
                                                     request.toJson());

        return result.isHttpOk();
    } catch (const std::exception& e) {
        std::cerr << "Exception in send operation: " << e.what() << std::endl;
        return false;
    }
}

bool MessagingChannelApi::disconnect(const std::string& sessionId) {
    try {
        udpClient_->close();
    } catch (const std::exception& e) {
        std::cerr << "Error closing UDP client: " << e.what() << std::endl;
    }

    try {
        SessionRequest request(sessionId);
        HttpClientResult result = httpClient_->post(getActionUrl("disconnect"),
                                                     request.toJson());

        httpClient_->closeAll();

        return result.isHttpOk();
    } catch (const std::exception& e) {
        std::cerr << "Exception in disconnect operation: " << e.what() << std::endl;
        return false;
    }
}

bool MessagingChannelApi::udpPush(const std::string& message,
                                  const std::string& destination,
                                  const std::string& sessionId) {
    try {
        EventMessageRequest request;
        request.sessionId = sessionId;
        request.type = EventType::CHAT_TEXT;
        request.to = destination;
        request.content = message;
        request.encrypted = false;

        UdpEnvelope envelope("push", request.toJson());

        return udpClient_->send(envelope);
    } catch (const std::exception& e) {
        std::cerr << "Exception in udpPush operation: " << e.what() << std::endl;
        return false;
    }
}

EventMessageResult MessagingChannelApi::udpPull(const std::string& sessionId,
                                               const ReceiveConfig& config) {
    EventMessageResult result;

    try {
        MessageReceiveRequest request;
        request.sessionId = sessionId;

        // Apply default poll source if not specified in config
        ReceiveConfig effectiveConfig = config;
        if (effectiveConfig.pollSource.empty()) {
            effectiveConfig.pollSource = defaultPollSource_;
        }
        request.receiveConfig = effectiveConfig;

        UdpEnvelope envelope("pull", request.toJson());

        json response = udpClient_->sendAndWait(envelope, 3000);

        if (!response.is_null()) {
            if (response.contains("status") && response["status"] == "ok") {
                if (response.contains("result")) {
                    json resultJson = response["result"];
                    if (resultJson.contains("status") && resultJson["status"] == "success") {
                        if (resultJson.contains("data")) {
                            return EventMessageResult::fromJson(resultJson["data"]);
                        }
                    }
                }
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in udpPull operation: " << e.what() << std::endl;
    }

    return result;
}

std::map<std::string, std::string> MessagingChannelApi::createAgentMetadata() const {
    std::map<std::string, std::string> metadata;
    metadata["agentType"] = "CPP-AGENT";
    metadata["descriptor"] = "hmdev::messaging::MessagingChannelApi";
    return metadata;
}

std::string MessagingChannelApi::getActionUrl(const std::string& action) const {
    return "/" + action;
}

} // namespace messaging
} // namespace hmdev

